### PR TITLE
FEATURE: Add site setting and wizard step to set base font

### PR DESCRIPTION
**Pull request by nbianca**
This commit builds on Neil's work, who added a site setting to control the base font, and adds a wizard step to set the base font. It looks like this:

image

### CODE REVIEW

**Comment by eviltrout** [comment]
This is looking nice. I don't see anything wrong with the code but I have a couple comments:

- It looks like these fonts are awfully similar. Maybe my eyesight is bad but I can't really tell the difference?
- Are there issues with CSP for adding a style tag like this? cc @pmusaraj

**Comment by pmusaraj** [comment]
No, I don't think there are any CSP issues with inline style tags (only inline JS).

**Comment by hnb-ku** [comment]
I think we can skip the `ttf` font files since `woff2` is [supported](https://caniuse.com/#feat=woff2) by all the browsers we support

File: config/locales/server.en.yml [hunk 1]
```diff
@@ -4677,6 +4679,10 @@ en:
           </ul>
           Popular theme components (for more, browse <a href='https://meta.discourse.org/c/theme/l/latest' target='_blank'>#theme</a>)"
 
+      fonts:
+        title: "Fonts"
+        description: "The base font used in most places of your site. This can be changed later and themes can override it."
```
Comment (eviltrout) [inline config/locales/server.en.yml:4684]:
```suggestion
        description: "Choose a font that matches your theme or branding."
```

Comment (eviltrout) [reply config/locales/server.en.yml:4684]: My thinking here is that we don't tell them that other settings can be changed (they all can.) And themes seems a bit advanced at this point.

File: config/locales/server.en.yml [hunk 2]
```diff
@@ -2224,6 +2224,8 @@ en:
     push_notifications_prompt: "Display user consent prompt."
     push_notifications_icon: "The badge icon that appears in the notification corner. A 96×96 monochromatic PNG with transparency is recommended."
 
+    base_font: "Font to use in most places on the site. Themes can override."
```
Comment (eviltrout) [inline config/locales/server.en.yml:2227]: Is this still used?

Comment (nbianca) [reply config/locales/server.en.yml:2227]: This is used for the site setting.

File: config/locales/server.en.yml (hunk 1 above)
Comment (nbianca) [reply config/locales/server.en.yml]: Yes, I agree with you. I decided to remove this message completely and leave only the font previews, similar to how we do it on the themes page. 😊

**Comment by SamSaffron** [comment]
looks like we need a rebase @nbianca 

This is adding 10s of megs to our git repo, perhaps we download these fonts on demand? I know this will complicate things but I do worry about adding lots of unused assets to our main repo. 

Regarding inline style, I don't think there are CSP concerns but I do much prefer keeping styling out of HTML if we can it is a lot harder to reason about things. 

One thing we could possibly do here is just maintain a bunch theme components and have the UI simply glue up the theme component to switch the font, that would keep this extension super clean and we would not have to change core that much only wizard.

**Comment by davidtaylorhq** [comment]
@SamSaffron - @nbianca and I chatted about this. I'm worried that trying to automatically install/update/attach/detach theme components using the wizard will be quite risky. There are so many edge cases to think about. (e.g. what if the component is already installed on a non-default theme? What if it's disabled? What if it's not up to date? How/when do we update it? etc.)

Also, if we do this with remote theme components, we will have to scrap the live preview @nbianca added to the wizard (because we won't have access to the font files).

I notice that the bulk of the file size comes from just two fonts (4mb each). We could scrap those options and save a lot of space.

Or, if we really wanted to keep these fonts out of the core repo, maybe we could package them in a Gem? (Kinda like how ember works at the moment)

File: Gemfile [hunk 3]
```diff
@@ -68,6 +68,7 @@ gem 'http_accept_language', require: false
 # behavior, we will push these versions up when upgrading ember
 gem 'ember-rails', '0.18.5'
 gem 'discourse-ember-source', '~> 3.12.2'
+gem 'discourse-fonts', git: 'https://github.com/discourse/discourse-fonts.git'
```
Comment (ZogStriP) [inline Gemfile:71]: Can you publish that gem instead of using git directly?

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -1,5 +1,5 @@
 <!DOCTYPE html>
-<html lang="<%= html_lang %>" class="<%= html_classes %>">
+<html lang="<%= html_lang %>" class="<%= html_classes %>" style="--font-setting: <%= SiteSetting.base_font_stack %>">
```
Comment (ZogStriP) [inline app/views/layouts/application.html.erb:2]: Why do you need to put this style inline?

File: public/fonts [hunk 5]
```diff
@@ -0,0 +1 @@
+/home/bianca/Work/discourse-fonts/vendor/assets/fonts
```
Comment (ZogStriP) [inline public/fonts:1]: Did you mean to commit that file?

**Review by ZogStriP — CHANGES_REQUESTED** [review]
It's a bit tight on the test side. Can you add an integration test that go through the wizard and ensure the font was properly changed?

File: Gemfile (hunk 3 above)
Comment (nbianca) [reply Gemfile:71]: Sure, I can do that.

File: config/application.rb [hunk 6]
```diff
@@ -307,6 +309,32 @@ def config.database_configuration
...
+                font-family: #{font[:name]};
+                src: asset-url("/fonts/#{variant[:filename]}") format("#{variant[:format]}");
+                font-weight: #{variant[:weight]};
+              }
+            EOF
+          end
+        end
+      end
+    end
+
```
Comment (SamSaffron) [inline config/application.rb:337]: we need to improve this, we can not do this every time we boot. we could do this as a custom step on our scss compiler or something like that. writing a scss file every time the app boots is very fragile.

**Pull request merged**